File: solutions/go/interest-is-interesting/1/interest_is_interesting.go
package interest

// InterestRate returns the interest rate for the provided balance.
func InterestRate(balance float64) float32 {
	if balance < 0.0 {
		return 3.213
	} else if balance < 1000.0 {
		return 0.5
	} else if balance < 5000.0 {
		return 1.621
	} else {
		return 2.475
	}
}

func Interest(balance float64) float64 {
	return balance * float64(InterestRate(balance)) * 0.01
}

func AnnualBalanceUpdate(balance float64) float64 {
	return balance + balance*(float64(InterestRate(balance))*0.01)
}

func YearsBeforeDesiredBalance(balance, targetBalance float64) int {
	i := 0
	for i = 0; true; i++ {
		if balance >= targetBalance {
			break
		} else {
			balance = AnnualBalanceUpdate(balance)
		}
	}
	return i
}

File: solutions/cpp/interest-is-interesting/1/interest_is_interesting.cpp
double interest_rate(double balance) {
    if (balance < 0.0)
        return 3.213;
    else if (balance < 1000.0)
        return 0.5;
    else if (balance < 5000.0)
        return 1.621;
    else 
        return 2.475;
}

double yearly_interest(double balance) {
    return balance * interest_rate(balance) * 0.01;
}

double annual_balance_update(double balance) {
    return balance + yearly_interest(balance);
}

int years_until_desired_balance(double balance, double target_balance) {
    int i{};
    while (true) { 
        if (balance < target_balance) 
            balance = annual_balance_update(balance);
        else 
            break;
        i++;
    }
    return i;
}
